[Proposal] remove lock from SocketAsyncEngine

adamsitnik opened the pull request:
While working on #36019 I've noticed that we could simplify the `SocketAsyncEngine` by doing two things:

* create the engines up-front and avoid locks when getting engine for a given context. It's a clear win for the majority of the cases (having a single engine), but an extra cost for scenarios with multiple (`n`) engines: the first async call is going to create `n` engines (and `n` epolls)
* use existing socket file descriptors, don't generate our own ids using `IntPtr`s, and avoid lock when adding new context to the engine. 

I don't know why we are generating new ids instead of using existing socket file descriptors (there was probably a good reason for that), so this is just a proposal.

/cc @tmds @stephentoub

adamsitnik committed:
remove lock from SocketAsyncEngine

stephentoub commented:
> I don't know why we are generating new ids instead of using existing socket file descriptors (there was probably a good reason for that), so this is just a proposal.

https://github.com/dotnet/runtime/commit/5e64fa3322294d2440c722cfbb2993b075aa05b0

tmds commented:
> I don't know why we are generating new ids instead of using existing socket file descriptors (there was probably a good reason for that), so this is just a proposal.

I was wondering about that too after https://github.com/dotnet/runtime/pull/36019.

I was chatting with @stephentoub about it, and I think another way to resolve 5e64fa3 would be to check socket readability before reading SO_ERROR for connect completion.

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -289,6 +169,24 @@ private SocketAsyncEngine()
 
+        private void AddToMap(SocketAsyncContext context)
+        {
+            IntPtr socketFileDescriptor = context.GetSocketFileDescriptor();
```
This is using `DangerousGetHandle` which may cause issues (besides 5e64fa3) due to fd recycling.

adamsitnik committed:
dont try to Free a token that was not allocated

adamsitnik commented:
@stephentoub @tmds thanks for providing the additional context.

This makes me wonder why simply don't remove given socket from `epoll` when it's being closed?

To my understanding we are just adding to `epoll`, never actually removing anything from it.

adamsitnik committed:
remove the socket from epoll

tmds commented:
> This makes me wonder why simply don't remove given socket from epoll when it's being closed?

That happens automatically in the kernel. The race happens when the `fd` is already in the `events` array returned by `epoll_wait`. When we look it up in the `ConcurrentDictionary`, we don't know it was meant for the previous/current `SocketAsyncContext` that used/uses the `fd`.

adamsitnik committed:
make Token a class to make sure that it can never be created with a null engine

remove from dictionary first, then remove from epoll

tmds commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs:
```diff
@@ -367,6 +373,14 @@ private bool TryRegister(SafeSocketHandle socket, out Interop.Error error)
+            error = Interop.Sys.TryChangeSocketEventRegistration(_port, socket,
+                Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write,
+                Interop.Sys.SocketEvents.None, socket.DangerousGetHandle());
+            return error == Interop.Error.SUCCESS;
```
This happens automagically in the kernel when the socket is closed. And it doesn't solve the race issue between the epoll thread and the thread that creates/closes sockets. See https://github.com/dotnet/runtime/pull/36115#issuecomment-626569425.

adamsitnik committed:
add context to the map first to make sure that the very first epoll|kqueue notification can be handled as soon as we receive it

if registration fails, remove context from the map

remove context from the map to make sure that we stop handling notifications first

adamsitnik committed:
minor refactor

adamsitnik commented:
@tmds thanks for all the great insights. I am going to close it for now and most probably send a simplified version later this week (just removing lock, not removing the id system)
